src/mrpro/operators/_CartesianSamplingOp.py

```diff
@@ -0,0 +1,144 @@
+        ktraj_tensor = traj.as_tensor()
+
+        if traj_type_kzyx[-1] == TrajType.ONGRID:  # kx
+            kx_idx = ktraj_tensor[-1, ...] + encoding_shape.x // 2
+        else:
+            encoding_shape.x = ktraj_tensor.shape[-1]
+            kx_idx = rearrange(torch.arange(ktraj_tensor.shape[-1]), 'kx->1 1 1 kx')
+
+        if traj_type_kzyx[-2] == TrajType.ONGRID:  # ky
+            ky_idx = ktraj_tensor[-2, ...] + encoding_shape.y // 2
+        else:
+            encoding_shape.y = ktraj_tensor.shape[-2]
+            ky_idx = rearrange(torch.arange(ktraj_tensor.shape[-2]), 'ky->1 1 ky 1')
+
+        if traj_type_kzyx[-3] == TrajType.ONGRID:  # kz
+            kz_idx = ktraj_tensor[-3, ...] + encoding_shape.z // 2
+        else:
+            encoding_shape.z = ktraj_tensor.shape[-3]
+            kz_idx = rearrange(torch.arange(ktraj_tensor.shape[-3]), 'kz->1 kz 1 1')
+
+        kidx = (
+            (kz_idx * encoding_shape.y * encoding_shape.x + ky_idx * encoding_shape.x + kx_idx)
+            .flatten()
+            .to(dtype=torch.int64)
+        )
+        self.register_buffer('_fft_idx', kidx.unsqueeze(0))
+        # we can skip the indexing if the data is already sorted
+        self._needs_indexing = not torch.all(torch.diff(kidx) == 1)
+
+        self._kshape = traj.broadcasted_shape
+        self._encoding_shape = encoding_shape
+
+    def forward(self, x: torch.Tensor) -> tuple[torch.Tensor,]:
+        """Forward operator which selects acquired k-space data from k-space.
+
+        Parameters
+        ----------
+        x
+            k-space with dimensions given by encoding_shape
+
+        Returns
+        -------
+            k-space data in original (i.e. acquired) dimensions
+        """
+        if self._encoding_shape != SpatialDimension(*x.shape[-3:]):
+            raise ValueError('k-space data shape missmatch')
+
+        if not self._needs_indexing:
+            return (x,)
+
+        x_2d = rearrange(x, '... k2 k1 k0 -> (...) (k2 k1 k0)')
+        # take_along_dim does broadcast
+        x_indexed = torch.take_along_dim(x_2d, self._fft_idx, dim=-1)
+        # reshape to (other1, other2 ..., k2, k1, k0)
+        x_reshaped = x_indexed.reshape(x.shape[:-3] + self._kshape[-3:])
+
+        return (x_reshaped,)
+
+    def adjoint(self, y: torch.Tensor) -> tuple[torch.Tensor,]:
+        """Adjoint operator sorting data into the encoding_space matrix.
+
+        Parameters
+        ----------
+        y
+            k-space data in original (i.e. acquired shape)
+
+        Returns
+        -------
+            k-space data sorted into encoding_space matrix
+        """
+
+        if self._kshape[-3:] != y.shape[-3:]:
+            raise ValueError('k-space data shape missmatch')
+
+        if not self._needs_indexing:
+            return (y,)
+
+        y_2d = rearrange(y, '... k2 k1 k0 -> (...) (k2 k1 k0)')
+
+        # scatter does not broadcast, so we need to expand the fft_idx to the full dataset
```

Comment:
Suggested additional comment: 
fft_idx is expanded by creating len(y_2d) copies of the indices. In the next step torch.scatter_ will then ensure that the index is correctly extended to the len(y_2d) dimensions.